File: algorithms/03_sorting/03_insertion_sort/python/insertion_sort.py
def insert(L, i):
    """ (list, int) -> NoneType

    Precondition: L[:i] is sorted from smallest to largest.

    Move L[i] to where it belongs in L[:i + 1].

    >>> L = [7, 3, 5, 2]
    >>> insert(L, 1)
    >>> L
    [3, 7, 5, 2]
    """

    # The value to be inserted into the sorted part of the list.
    value = L[i]

    # Find the index, j, where the value belongs.
    # Make room for the value by shifting.
    j = i
    while j != 0 and L[j - 1] > value:
        # Shift L[j - 1] one position to the right to L[j].
        L[j] = L[j - 1]
        j = j - 1

    # Put the value where it belongs.
    L[j] = value


def insertion_sort(L):
    """ (list) -> NoneType

    Sort the items of L from smallest to largest.

    >>> L = [7, 3, 5, 2]
    >>> insertion_sort(L)
    >>> L
    [2, 3, 5, 7]

    >>> L = [99, 44, 6, 2, 1, 5, 63, 87, 283, 4, 0]
    >>> insertion_sort(L)
    >>> L
    [0, 1, 2, 4, 5, 6, 44, 63, 87, 99, 283]
    """

    for i in range(len(L)):
        insert(L, i)


if __name__ == "__main__":
    import doctest

    doctest.testmod()

File: algorithms/03_sorting/03_insertion_sort/js/insertion_sort.js
function swap(arr, i, j) {
  const temp = arr[i]
  arr[i] = arr[j]
  arr[j] = temp
}

/*
 * @i: the index of the value to be inserted
 */
function insert(arr, i) {
  // The value to be inserted
  const value = arr[i]
  // Grab the next unsorted element
  let currentIndex = i

  // To find the index where the value belongs, repeatedly compare it with the elements to its left, going right to left
  while (currentIndex > 0) {
    // If the value is less than the element to its left, swap the elements
    if (value < arr[currentIndex - 1]) {
      swap(arr, currentIndex, currentIndex - 1)
    }
    // Shift current index left
    currentIndex--
  }
}

export function insertionSort(array) {
  const arr = [...array]

  // Walk through the array
  for (let i = 0; i < array.length; i++) {
    // Call the insert function on each index
    insert(arr, i)
  }

  return arr
}
